0.6 Release Blog Post (#13)
GameDevTecnico/cubos-blog: content/0-6.rst

0.6 is out!
###########

:date: 2025-03-12 00:00:00
:category: Release
:summary: Cubos 0.6 Release!

.. role:: dim
    :class: m-text m-dim

New 0.6 Features
================

This blog post covers the most important changes, but you can check out the full changelog in our `repository <https://github.com/GameDevTecnico/cubos/blob/main/CHANGELOG.md>`_.

On the Engine
-------------

Physics with Warm-Starting :dim:`(@fallenatlas)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Continuing the work from the previous release, we finished **warm-starting** for our physics simulation. Warm-starting takes the information stored in the contact manifold between the two contacting bodies, particularly the total impulses applied over each point in the previous frame, and applies it before any solving is performed.

Reviewer: ```suggestion
Continuing the work from the previous release, we finished **warm-starting** for our physics simulation. Warm-starting uses the information stored in the contact manifold between two bodies, especially the total impulses applied to each contact point in the previous frame, and applies it before solving begins.
```

> It works based on the assumption that the impulses applied in the previous frame are a good estimate for the impulses that will also be applied in the current frame. Therefore, by applying them to the body before solving the collision constraint, the body should already be closer to its ideal position. Over time, these impulses will stabilize (since our solver is a "dirty" solver), leading to less jitter.

Reviewer: ```suggestion
It works based on the assumption that the impulses applied in the previous frame are a good estimate for the impulses that will also be applied in the current frame. Therefore, by applying them to the body before solving the collision constraint, the body should already be closer to its ideal position. Over time, these impulses will stabilize (since our solver is a "dirty" solver), resulting in less jitter.
```

> This works especially well for stacks of bodies since, in this case, the bodies will push each other away, creating instability. However, if they're already closer to their ideal position, this instability will be reduced. With warm-starting, we can stack more bodies, and when the stack isn't fully stable it stays together for longer.

Reviewer: ```suggestion
This is particularly effective for stacks of bodies, where the bodies push against each other, creating instability. However, if they're already closer to their ideal position, this instability will be reduced. With warm-starting, we can stack more bodies, and when the stack isn't fully stable, it stays together for longer.
```

to change shaders' source code at runtime. To solve this, we added new methods to the `Shader <https://docs.cubosengine.org/classcubos_1_1engine_1_1Shader.html>`_
class to prepend ``#define`` macros and recompile the shader.

.. code-block:: cpp

    auto vs = vsAsset->shaderStage(); // unmodified shader
    vs = vsAsset->builder().with("RENDER_PICKER").build(); // shader with #define RENDER_PICKER
    vs = vsAsset->builder().with("MAX_CSM_SPLITS", "5").build(); // shader with #define MAX_CSM_SPLITS 5

Unfortunately, the actual ability to toggle render picking didn't make it to this release, but it should make it to the next one!

On the Core
-----------

Added Frustum geometry and intersection utilities :dim:`(@mkuritsu)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A new geometry object was added, the **frustum**, and additionally, two new intersection functions to check if a `box <https://docs.cubosengine.org/structcubos_1_1core_1_1geom_1_1Box.html>`_ or a `capsule <https://docs.cubosengine.org/structcubos_1_1core_1_1geom_1_1Capsule.html>`_ intersects a frustum.

> This new geometry utility will prove useful to perform *frustum culling* in a future release, which could bring a great graphics performance improvement when rendering large scenes with **Cubos**.

Reviewer: ```suggestion
This new geometry utility will be useful for *frustum culling* in a future release, potentially improving graphics performance when rendering large scenes with **Cubos**.
```

New Demo Sneak Peek
===================

We have been working on a new game made with **Cubos**, and we're excited to share a sneak peek with you!
It's a competitive multiplayer game where you and your team pilot an airship and fight against other teams.

> We're planning to implement networked multiplayer, so that you can play with your friends online - this will be a first for **Cubos**!

Reviewer: ```suggestion
We plan to implement networked multiplayer, allowing you to play with friends online — this will be a first for **Cubos**!
```

> Expect to hear more about it after the 0.7 release blog post!

Reviewer: ```suggestion
Stay tuned for more details after the 0.7 release blog post!
```

.. image:: images/0.6/airships_sneak_peek.png

Next Steps
==========

In the next release, which should be out by the end of April, we're planning to work on the following features:

> * A menu bar and a sane default layout for **Tesseratos**.

Reviewer: ```suggestion
* A menu bar and a sensible default layout for **Tesseratos**.
```

* A proper scene editor for **Tesseratos**.
* Automated inertia computation for arbitrary voxel shapes.
* Frustum culling when rendering.
* Various other performance improvements in the renderer.
* Better integration of mouse input with the high-level input system.
* And many other new features and bug fixes!

> Until then, we'll also participate in the `Azul Jam <https://azulgamejam.com/>`_, a game jam that will take place over the weekend of March 22nd.

Reviewer: ```suggestion
Until then, we'll also participate in the `Azul Jam <https://azulgamejam.com/>`_, a game jam that will take place over the weekend of March 22nd.
```
```suggestion
In the meantime, we'll also be participating in the `Azul Jam <https://azulgamejam.com/>`_, a game jam happening over the weekend of March 22nd.
```

Reviewer: In the meantime, we'll also be participating in the `Azul Jam <https://azulgamejam.com/>`_, a game jam happening over the weekend of  March 22nd. ~~is the wrong date!!
 March 28<sup>th</sup>.
